clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java

```diff
@@ -64,32 +84,41 @@ public int position() {
         return buffer.position();
     }
 
-    public int capacity() {
-        return buffer.capacity();
+    public int remaining() {
+        return buffer.remaining();
     }
 
     public int limit() {
         return buffer.limit();
     }
 
     public void position(int position) {
-        if (position > buffer.limit())
-            expandBuffer(position);
+        maybeExpandBuffer(position - buffer.position());
         buffer.position(position);
     }
 
-    private void expandBuffer(int size) {
-        int expandSize = Math.max((int) (buffer.capacity() * REALLOCATION_FACTOR), size);
-        ByteBuffer temp = ByteBuffer.allocate(expandSize);
-        if (buffer.hasArray()) {
-            temp.put(buffer.array(), buffer.arrayOffset(), buffer.position());
-        } else {
-            int limit = buffer.position();
-            for (int i = 0; i < limit; i++)
-                temp.put(buffer.get(i));
-        }
+    /**
+     * The capacity of the first internal ByteBuffer used by this class. This is useful in cases where a pooled
+     * ByteBuffer was passed via the constructor and it needs to be returned to the pool.
```

Comment:
Yeah, it seems the buffer pool can handle an expanded buffer. Also looking back, this behavior was changed in 0.10.2. Let's keep it as is since it's a bit simpler and hopefully this case is rare in practice.